### DISCUSSION TITLE
gridmap isnt good enough.

### DISCUSSION THREAD

**Discussion by BaddRadish** [godotengine/godot-proposals#5226] [3D]
im grateful for Gridmap, but that said as it stands right now, gridmap is behind other aspects of godot*.

first off, we need better tile drawing tools.
+ a minecraft like draw, where we cast a ray and draw on the face of the first thing we hit. this is obviously a good idea since most blocks are placed next to other blocks, with a few exceptions. fall back to drawing where the ray hits the grid when we dont hit anything. but we probably need to be able to override this for more careful control and for floating things. thats fine hold shift and click or something. maybe a toggle swaps those rolls? or enables/disables the minecraft style. you know for people who prefer the current system, or need the old tool breifly.
+ a transparent draw cursor so we can see what we're doing. maybe it changes color to indicate if we're going to overwrite vs going to add vs holding down add/overwrite vs holding down erase.
+ the ability to draw / fill in rectangular prisms.
+ flood fill along a plane. useful for floors and walls.
+ rectangular prisms of cell selection for cut/copy/paste / drag and drop / whatever. drag and drop should work with the grid obviously.
+ arbitrary cell selection.
+ drag and clone a selection, useful for drawing volumes for hills and such in a minecraft style level

also, the grid being actually aligned with where we're drawing would be useful. i suggest we draw a 1 tall plane of boxes that we're placing /into/ instead of a plane of squares.

Furthermore, what we /really/ need is 3d tilemap. Everything useful for tiling in 2d is useful for tiling in 3d. as you probably know a lot of people have already been discussing this on here. when we go this route i suggest we rename it TileMap3D or change tilemap to GridMap2D to really tie those two systems together conceptually. as much as possible the interface should be the same so devs can migrate easily.

thats all i got for now. hopefully this sparks a discussion as to how we can bring this tool up to godot's other standards for quality of life. thanks for your time.

ps i think we need at least /some/ improvements to gridmap before gd4. its really one of the wonkier bits in godot.

**Comment by Calinou** [comment]
There are already several proposals on the subject: https://github.com/godotengine/godot-proposals/issues/1453, https://github.com/godotengine/godot-proposals/issues/1715, https://github.com/godotengine/godot-proposals/issues/1713, https://github.com/godotengine/godot-proposals/issues/3710, https://github.com/godotengine/godot-proposals/issues/114, https://github.com/godotengine/godot-proposals/issues/3518

> ps i think we need at least /some/ improvements to gridmap before gd4. its really one of the wonkier bits in godot.

In an effort to release it sooner than later, Godot 4.0 has already entered feature freeze phase. As such, no major rework of GridMap is planned for 4.0 (but minor enhancements may be accepted).

> gridmap has an interface built by an engineer. it needs to be shot and burned and replaced by something made for humans.

I have to remind you that we have a [Code of Conduct](https://godotengine.org/code-of-conduct). Please stay constructive.

**Reply by BaddRadish** [reply]
a misguided attempt to inject a little humor. my bad. i adjusted the post.
